### PR TITLE
[BugFix] fix the issue of inaccurate limit data after show export sorting

### PR DESCRIPTION
**Pull request by LinYuHeng-Con**
Why I'm doing:
    Inaccurate limit data after show export sorting
    show export from alg order by jobId desc
    image
    show export from alg order by jobId desc limit 20
    image
What I'm doing:
    Sort first and then limit

Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by 873253814** [comment]
This is my first time submitting code, and the process has been under approval. How can I keep the workflow running

**Comment by jaogoy** [comment]
@873253814 I'm not clear about the error, could you point it out for my?
I think, the result you list is OK both for the sort and sort & limit. (maybe there is a bug about sort and limit, I don't)

**Comment by 873253814** [comment]
> @873253814 I'm not clear about the error, could you point it out for my? I think, the result you list is OK both for the sort and sort & limit. (maybe there is a bug about sort and limit, I don't)

you can take a look at the comparison of the results from the two pictures above

After executing the statement 'show export from alg order by jobId desc limit 20', 
the normal return result should be:
1701248288510_4648EA72-86DD-456a-8330-9F20BA4AE1E3

But in reality, the following values were returned:
image

Missing data such as 299958626,299949781

**Comment by 873253814** [comment]
> @873253814 I'm not clear about the error, could you point it out for my? I think, the result you list is OK both for the sort and sort & limit. (maybe there is a bug about sort and limit, I don't)

https://forum.mirrorship.cn/t/topic/4023

**Comment by 873253814** [comment]
The code I modified this time has nothing to do with the exceptions in the workflow, why did the detection fail.

**Comment by 873253814** [comment]
Is it due to incomplete statements? But I haven't written this statement before
image

**Comment by 873253814** [comment]
> @873253814 I'm not clear about the error, could you point it out for my? I think, the result you list is OK both for the sort and sort & limit. (maybe there is a bug about sort and limit, I don't)

May I ask if you can help me take a look at the workflow issue? Can the workflow of AdminTest be restarted? The code I submitted does not match the exception thrown by the workflow. Thank you very much

**Comment by stdpain** [comment]
@gengjun-git

**Review by kangkaisen — DISMISSED** [review]

**Comment by kevincai** [comment]
@mergifyio rebase

File: fe/fe-core/src/main/java/com/starrocks/load/ExportMgr.java [hunk 1]
```diff
@@ -217,20 +219,16 @@ public List<List<String>> getExportJobInfosByIdOrState(
                 }
 
                 // filter job
-                if (jobId != 0) {
-                    if (id != jobId) {
-                        continue;
-                    }
+                if (jobId != 0 && (id != jobId)) {
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/load/ExportMgr.java:222]:
```suggestion
                if (jobId != 0 && id != jobId) {
```

File: fe/fe-core/src/main/java/com/starrocks/load/ExportMgr.java [hunk 2]
```diff
@@ -217,20 +219,16 @@ public List<List<String>> getExportJobInfosByIdOrState(
...
+                    continue;
                 }
 
-                if (states != null) {
-                    if (!states.contains(state)) {
-                        continue;
-                    }
+                if (states != null && (!states.contains(state))) {
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/load/ExportMgr.java:226]:
```suggestion
                if (states != null && !states.contains(state)) {
```

File: fe/fe-core/src/main/java/com/starrocks/load/ExportMgr.java [hunk 3]
```diff
@@ -217,20 +219,16 @@ public List<List<String>> getExportJobInfosByIdOrState(
...
+                    continue;
                 }
 
                 UUID jobQueryId = job.getQueryId();
-                if (queryId != null && (jobQueryId == null || !queryId.equals(jobQueryId))) {
+                if (queryId != null && (!queryId.equals(jobQueryId))) {
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/load/ExportMgr.java:231]:
```suggestion
                if (queryId != null && !queryId.equals(jobQueryId)) {
```

File: fe/fe-core/src/main/java/com/starrocks/load/ExportMgr.java [hunk 4]
```diff
@@ -305,31 +303,32 @@ public List<List<String>> getExportJobInfosByIdOrState(
...
         }
-        Collections.sort(exportJobInfos, comparator);
+        exportJobInfos.sort(comparator);
 
         List<List<String>> results = Lists.newArrayList();
-        for (List<Comparable> list : exportJobInfos) {
-            results.add(list.stream().map(e -> e.toString()).collect(Collectors.toList()));
+        for (int i = 0; i < exportJobInfos.size(); i++) {
+            results.add(exportJobInfos.get(i).stream().map(Object::toString).collect(Collectors.toList()));
+            if (i > resultNum - 1) {
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/load/ExportMgr.java:328]: looks like this break condition is not correct, will result one more record in the result list.

**Comment by kevincai** [comment]
@LinYuHeng-Con please fix the new compiling errors after code rebased to latest main branch.

**Comment by LinYuHeng-Con** [comment]
> @LinYuHeng-Con please fix the new compiling errors after code rebased to latest main branch.

image
I have resynchronized the main code to my branch. Is this enough?

**Comment by kevincai** [comment]
> > @LinYuHeng-Con please fix the new compiling errors after code rebased to latest main branch.
> 
> image I have resynchronized the main code to my branch. Is this enough?

I don't think so, check the failure logs: https://github.com/StarRocks/starrocks/actions/runs/9868902318/job/27251940532?pr=35854

File: fe/fe-core/src/main/java/com/starrocks/load/ExportMgr.java (hunk 4 above)
Comment (LinYuHeng-Con) [reply fe/fe-core/src/main/java/com/starrocks/load/ExportMgr.java:328]: I will fix it

**Comment by LinYuHeng-Con** [comment]
> > > @LinYuHeng-Con please fix the new compiling errors after code rebased to latest main branch.
> > 
> > 
> > image I have resynchronized the main code to my branch. Is this enough?
> 
> I don't think so, check the failure logs: https://github.com/StarRocks/starrocks/actions/runs/9868902318/job/27251940532?pr=35854

I will fix Test Class and recommit, Thank you very much

**Comment by LinYuHeng-Con** [comment]
> > > @LinYuHeng-Con please fix the new compiling errors after code rebased to latest main branch.
> > 
> > 
> > image I have resynchronized the main code to my branch. Is this enough?
> 
> I don't think so, check the failure logs: https://github.com/StarRocks/starrocks/actions/runs/9868902318/job/27251940532?pr=35854

I have completed the modifications. Could you please take a look again？

**Comment by kevincai** [comment]
> > > > @LinYuHeng-Con please fix the new compiling errors after code rebased to latest main branch.
> > > 
> > > 
> > > image I have resynchronized the main code to my branch. Is this enough?
> > 
> > 
> > I don't think so, check the failure logs: https://github.com/StarRocks/starrocks/actions/runs/9868902318/job/27251940532?pr=35854
> 
> I have completed the modifications. Could you please take a look again？

did you fix the unit test. I see the unit test gets removed.

**Comment by LinYuHeng-Con** [comment]
> > > > > @LinYuHeng-Con please fix the new compiling errors after code rebased to latest main branch.
> > > > 
> > > > 
> > > > image I have resynchronized the main code to my branch. Is this enough?
> > > 
> > > 
> > > I don't think so, check the failure logs: https://github.com/StarRocks/starrocks/actions/runs/9868902318/job/27251940532?pr=35854
> > 
> > 
> > I have completed the modifications. Could you please take a look again？
> 
> did you fix the unit test. I see the unit test gets removed.

Yes, this unit test was originally written by me for this modification. After testing and finding no issues, I deleted it

**Comment by kevincai** [comment]
shall keep the test cases as a guardian to the code quality.

**Comment by LinYuHeng-Con** [comment]
> shall keep the test cases as a guardian to the code quality.

I have added test cases, can you take a look at my commit again？

File: fe/fe-core/src/main/java/com/starrocks/load/ExportMgr.java [hunk 5]
```diff
@@ -305,31 +303,32 @@ public List<List<String>> getExportJobInfosByIdOrState(
...
-        Collections.sort(exportJobInfos, comparator);
+        exportJobInfos.sort(comparator);
 
         List<List<String>> results = Lists.newArrayList();
-        for (List<Comparable> list : exportJobInfos) {
-            results.add(list.stream().map(e -> e.toString()).collect(Collectors.toList()));
+        for (int i = 0; i < exportJobInfos.size(); i++) {
+            if (i > resultNum - 1) {
+                break;
+            }
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/load/ExportMgr.java:329]:
```suggestion
        int upperBound = Math.min(resultNum, exportJobInfos.size());
        for (int i = 0; i < upperBound; i++) {
```

**Review by kevincai — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/load/ExportMgr.java (hunk 5 above)
Comment (LinYuHeng-Con) [reply fe/fe-core/src/main/java/com/starrocks/load/ExportMgr.java]: Thank you for your review. I have made the necessary changes

**Review by wyb — APPROVED** [review]

**Comment by kevincai** [comment]
@mergifyio backport branch-3.3

**Comment by kevincai** [comment]
give up backport to branch-3.0 and branch-2.5

**Pull request merged**